doi/retro_doi_builder.py

```diff
@@ -0,0 +1,467 @@
+#!/usr/bin/env python
+"""
+Retroactively generate DOIs for raw and parsed data represented in uFrame.
+This script creates Raw File Set, Parsed Data Set, Parser, and Sensor DOIs.
+DOIDaemon logic will handle creating Array, Subsite, Node, and Cruise DOIs.
+
+Usage:
+    retro_doi_builder.py <uframe_ip> <cassandra_ip_list>...
+"""
+
+import re
+import csv
+import sqlite3
+import json
+import os
+import sys
+import requests
+import cassandra.cluster
+import operator
+from datetime import datetime
+import docopt
+import traceback
+import time
+import multiprocessing
+
+# Add parent directory to python path to locate the
+# metadata_service_api package
+sys.path.insert(0, '/'.join(
+        (os.path.dirname(os.path.realpath(__file__)), '..')))
+from metadata_service_api import MetadataServiceAPI
+from xasset_service_api.api import XAssetServiceAPI
+from doi_service_api.api import DOIServiceAPI, DOIServiceException
+
+RUTGERS_FILE_SERVER = 'https://rawdata.oceanobservatories.org:13580/'
+STREAMING_SENSOR_INVENTORY_SERVICE_TEMPLATE = 'http://{0}:12576/sensor/inv'
+
+NTP_UNIX_DELTA_SECONDS = 2208988800
+MILLIS_PER_DAY = 24 * 60 * 60 * 1000
+
+DATASET_L0_PROVENANCE_COLUMNS = [
+        'subsite', 'node', 'sensor', 'method', 'deployment', 'id', 'fileName',
+        'parserName', 'parserVersion']
+ALL_DATASET_L0_PROVENANCE = 'SELECT {0} FROM dataset_l0_provenance'.format(
+        ','.join(DATASET_L0_PROVENANCE_COLUMNS))
+
+STREAM_METADATA_SERVICE_URL_TEMPLATE = 'http://{0}:12571/streamMetadata'
+PARTITION_METADATA_SERVICE_URL_TEMPLATE = 'http://{0}:12571/partitionMetadata'
+XASSET_SERVICE_URL_TEMPLATE = 'http://{0}:12587/'
+DOI_SERVICE_URL_TEMPLATE = 'http://{0}:12588/doi'
+PARSED_URL_TEMPLATE = """%s/sensor/inv/%s/%s/%s/%s/%s?beginDT=%s&endDT=%s&
+                      execDPA=false&user=doilookup"""
+DATE_REGEX = re.compile("""\\d{4}[_-]?(0[1-9]|1[0-2])[-_]?(0[1-9]|[12][0-9]|
+                        3[01])""")
+
+
+# Take a timestamp representing milliseconds since the UNIX epoch and return
```

Comment:
It would be nice to have the time utilities in a central place (e.g. ooi-tools/common). Also, consider `java_time_to_iso8601` instead of `convertMillisToIso8601`